### ISSUE TITLE
TypeScript: pipeStdout (and other pipeXxx) is optional property 

### ISSUE THREAD

**Issue by koshic** [sindresorhus/execa#580]
What is the reason to keep it optional?

```ts
pipeStdout?<Target extends ExecaChildPromise<StdoutStderrAll>>(target: Target): Target;
pipeStdout?(target: WritableStream | string): ExecaChildProcess<StdoutStderrType>;
```

It can't be user without extra check or '!`:
<img width="667" alt="image" src="https://github.com/sindresorhus/execa/assets/20106607/f52eb6b3-43ed-4631-b69f-896ae0097bcc">

**Comment by ehmicky** [comment]
Hi @koshic,

`pipeStdout()` is `undefined` if the `stdout` option is `'ignore'`, `'ipc'`, a file descriptor integer, or a stream.
The same applies to `pipeStderr()`, but using the `stderr` option.

`pipeAll()` is `undefined` if either: 
  - Both the `stdout` and `stderr` options are `'ignore'`, `'ipc'`, a file descriptor integer, or a stream.
  - The `all` option is not `true`.

https://github.com/sindresorhus/execa/blob/834e3726f0c7b7f23827b6e09b69b66dedbddd47/lib/pipe.js#L31-L33

Ideally the types would take this into account. Would you be interested in submitting a PR to fix them?

**Comment by koshic** [comment]
@ehmicky yeah, let me experiment a bit )

**Comment by koshic** [comment]
As you can see, if we will use current strategy - modify options via generic parameter, and provide all possible combinations as execa definitions, we have to multiply all definitions by 3 (new modifiers: stdout + stdin + all) which makes code unmaintainable:

<img width="606" alt="image" src="https://github.com/sindresorhus/execa/assets/20106607/86723f46-3aa1-49ba-9d13-b6e6f78d0a68">

I see only one way - keep modifiers for options, but produce exact ExecaChildProcess type by additional conditional type wrapper which will work like switch-case statement:
```ts
O extends { stdin: 'pipe' /* + default value check */} ? { pipeStdout: () => {} } : { }
```

@ehmicky what do you think about it?

**Comment by ehmicky** [comment]
Yes, I think what you're highlighting was the reason I did not implement stricter typing in the first place when adding those pipe methods.

I agree with you. I'm not sure if this is what you are describing, but instead of method overloading, type inference should be used. I.e. only once `execa()` definition where `Options<EncodingType>` is being inferred from the parameters passed as input. Something like:

```ts
export function execa<EncodingType extends EncodingOption = DefaultEncodingOption>(
	file: string,
	arguments?: readonly string[],
	options?: Options<EncodingType> // EncodingType is now inferred based on the `options` passed as input
): ExecaChildProcess<EncodingType extends BufferEncodingOption ? Buffer : string>; // Which is used to guess whether `result.stdout|stderr|all` is a `string` or a `Buffer`
```

What do you think?

**Comment by koshic** [comment]
Yes, I mean the same pattern - define function signature once, and use smart generic types. So, I propose to split that task into 2:

First PR - to replace overloading with type inference. Not a breaking change, I believe.
Second PR - to extend generics with PipeXxx options. May be a breaking change due to pipeStdout will be not present in resulting type in some cases.

**Comment by ehmicky** [comment]
This sounds like a good plan!

I actually believe that this might not be a breaking change, because the type will go from `Function | undefined` to either `Function` or `undefined`, depending on the option being passed, so it is actually stricter. That being said, I am planning on making an unrelated PR which will require a new major release, so we can always bundle it with it, to be safe.

Thanks for your help!

**Comment by ehmicky** [comment]
@koshic Would you be interested in doing the second PR discussed above?

**Comment by koshic** [comment]
@ehmicky yeah, I need a few days to recover from my illness.

**Comment by ehmicky** [comment]
Thanks @koshic. Take care! :woman_health_worker:

**Comment by ehmicky** [comment]
This has been just been fixed in Execa 9.0.0, with the new `subprocess.pipe()` method. Please see (and share!) the [release post](https://medium.com/@ehmicky/execa-9-release-d0d5daaa097f) and the [changelog](https://github.com/sindresorhus/execa/releases/tag/v9.0.0).
